google/cloud/spanner/client.cc

```diff
@@ -99,16 +99,68 @@ StatusOr<std::int64_t> Client::ExecutePartitionedDml(
   return Status(StatusCode::kUnimplemented, "not implemented");
 }
 
-StatusOr<CommitResult> Client::Commit(
-    Transaction const& /*transaction*/,
-    std::vector<Mutation> const& /*mutations*/) {
-  return Status(StatusCode::kUnimplemented, "not implemented");
+StatusOr<CommitResult> Client::Commit(Transaction const& transaction,
+                                      std::vector<Mutation> const& mutations) {
+  auto session = GetSession();
+  if (!session) {
+    return std::move(session).status();
+  }
+  grpc::ClientContext context;
+  google::spanner::v1::CommitRequest request;
+  request.set_session(session->session_name());
+  for (auto const& m : mutations) {
+    *request.add_mutations() = m.as_proto();
+  }
+  using F = std::function<int(spanner_proto::TransactionSelector&)>;
+  internal::Visit(
+      transaction, F([&request](spanner_proto::TransactionSelector& s) {
+        if (!s.id().empty()) {
+          request.set_transaction_id(s.id());
+        } else if (s.has_begin()) {
+          *request.mutable_single_use_transaction() = s.begin();
+        } else if (s.has_single_use()) {
+          *request.mutable_single_use_transaction() = s.single_use();
+        }
+        // TODO(#306) - Make Visit() work with functions that return `void`.
```

Comment:
Typically you would need to do the `Commit()` in here, at least because in the `has_begin()` case you also need to `s.set_id()` with the returned transaction id.  And that means you typically want to return something as well.

Reply:
Can you take another look and see if I am using the `Transaction` class as you intended?